dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/imports/programrule/RuleEngineMapperTest.java

```diff
@@ -0,0 +1,392 @@
+        new RuleAttributeValue(trackedEntityAttribute.getUid(), SAMPLE_VALUE_A);
+
+    RuleEnrollment ruleEnrollment =
+        mapper.toRuleEnrollment(enrollment, List.of(trackedEntityAttributeValue), trackerPreheat);
+
+    assertEnrollment(enrollment, ruleEnrollment);
+  }
+
+  @Test
+  void shouldMapDBEnrollmentToRuleEnrollment() {
+    Enrollment enrollment = dbEnrollment();
+    RuleAttributeValue trackedEntityAttributeValue =
+        new RuleAttributeValue(trackedEntityAttribute.getUid(), SAMPLE_VALUE_A);
+
+    RuleEnrollment ruleEnrollment =
+        mapper.toRuleEnrollment(enrollment, List.of(trackedEntityAttributeValue));
+
+    assertEnrollment(enrollment, ruleEnrollment);
+  }
+
+  @Test
+  void shouldMapDBEnrollmentToRuleEnrollmentWhenOrganisationUnitCodeIsNull() {
+    OrganisationUnit organisationUnitWithNullCode = createOrganisationUnit('A');
+    organisationUnitWithNullCode.setCode(null);
+    Enrollment enrollment = dbEnrollment();
+    enrollment.setOrganisationUnit(organisationUnitWithNullCode);
+    RuleAttributeValue trackedEntityAttributeValue =
+        new RuleAttributeValue(trackedEntityAttribute.getUid(), SAMPLE_VALUE_A);
+
+    RuleEnrollment ruleEnrollment =
+        mapper.toRuleEnrollment(enrollment, List.of(trackedEntityAttributeValue));
+
+    assertEnrollment(enrollment, ruleEnrollment);
+  }
+
+  @Test
+  void shouldMapAttributes() {
+    Attribute textAttribute = attribute(trackedEntityAttribute, SAMPLE_VALUE_A);
+    Attribute nullTextAttribute = attribute(trackedEntityAttribute, null);
+    Attribute numericAttribute = attribute(numericTrackedEntityAttribute, "3");
+    Attribute nullNumericAttribute = attribute(numericTrackedEntityAttribute, null);
+    Attribute booleanAttribute = attribute(booleanTrackedEntityAttribute, "true");
+    Attribute nullBooleanAttribute = attribute(booleanTrackedEntityAttribute, null);
+
+    List<Attribute> attributes =
+        List.of(
+            textAttribute,
+            nullTextAttribute,
+            numericAttribute,
+            nullNumericAttribute,
+            booleanAttribute,
+            nullBooleanAttribute);
+
+    TrackerPreheat preheat = new TrackerPreheat();
+    preheat.put(
+        TrackerIdSchemeParam.UID,
+        List.of(
+            trackedEntityAttribute, numericTrackedEntityAttribute, booleanTrackedEntityAttribute));
+    List<RuleAttributeValue> ruleAttributes = mapper.toRuleAttributes(attributes, preheat);
+
+    assertAttributes(SAMPLE_VALUE_A, trackedEntityAttribute, ruleAttributes);
+    assertAttributes("", trackedEntityAttribute, ruleAttributes);
+    assertAttributes("3", numericTrackedEntityAttribute, ruleAttributes);
+    assertAttributes("0", numericTrackedEntityAttribute, ruleAttributes);
+    assertAttributes("true", booleanTrackedEntityAttribute, ruleAttributes);
+    assertAttributes("false", booleanTrackedEntityAttribute, ruleAttributes);
+  }
+
+  private void assertAttributes(
```

Comment:
RuleAttributeValue is https://github.com/dhis2/dhis2-rule-engine/blob/5fa1af76cf562d9c7ac59ce42ec28a22aae35154/src/commonMain/kotlin/org/hisp/dhis/rules/models/RuleAttributeValue.kt#L7-L8 right? Does this implement equals?

If so a better assertion would be to create an expected `RuleAttributeValue` and rely on that equals implementation. This allows you to write something like

```java
assertContainsOnly(List.of(
 new RuleAttributeValue("3", "uid"),
...
 ), ruleAttributes);
```

That means you can leverage existing assertions without having to do much. you also get better error messages. `assertEquals` used in `assertContainsOnly` gives you somewhat of a diff. You can remove all that filter logic that would just leave you scratching your head at first with `assertTrue` was false.

If you only want to assert on parts of the fields this does not work anymore of course. Here your type is a simple key-value pair.